src/stdune/sexp.ml

```diff
@@ -123,10 +127,32 @@ module Of_sexp = struct
     b ctx state
   let map t ~f = t >>| f
 
+  let get_user_context : type k. k context -> Univ_map.t = function
+    | Values (_, _, uc) -> uc
+    | Fields (_, _, uc) -> uc
+
+  let get key ctx state = (Univ_map.find (get_user_context ctx) key, state)
+
+  let set : type a b k. a Univ_map.Key.t -> a -> (b, k) parser -> (b, k) parser
+    = fun key v t ctx state ->
+      match ctx with
+      | Values (loc, cstr, uc) ->
+        t (Values (loc, cstr, Univ_map.add uc key v)) state
+      | Fields (loc, cstr, uc) ->
+        t (Fields (loc, cstr, Univ_map.add uc key v)) state
+
+  let set_many : type a k. Univ_map.t -> (a, k) parser -> (a, k) parser
+    = fun map t ctx state ->
+      match ctx with
+      | Values (loc, cstr, uc) ->
+        t (Values (loc, cstr, Univ_map.superpose uc map)) state
+      | Fields (loc, cstr, uc) ->
+        t (Fields (loc, cstr, Univ_map.superpose uc map)) state
```

Comment:
Is it worth it here to have a `map_context : 'k context -> (Univ_map.t -> Univ_map.t) -> 'k context'` here to make `set_many` and `set` a little simpler?

Reply:
This is a definition of the map_context function that I had in mind btw:

```ocaml
  let map_context
    : type k. k context -> f:(Univ_map.t -> Univ_map.t) -> k context
    = fun t ~f ->
      match t with
      | Values (loc, cstr, uc) -> Values (loc, cstr, f uc)
      | Fields (loc, cstr, uc) -> Fields (loc, cstr, f uc)
```

Reply:
yh that seems fine. It should be named `map_user_context` though, to avoid the confusion with the `context` type